rules/no-for-loop.js

```diff
@@ -335,7 +343,7 @@ const create = context => {
 					const shouldGenerateIndex = isIndexVariableUsedElsewhereInTheLoopBody(indexVariable, bodyScope, arrayIdentifierName);
 
 					const index = indexIdentifierName;
-					const element = elementIdentifierName || defaultElementName;
+					const element = elementIdentifierName || avoidCapture(defaultElementName, getChildScopesRecursive(bodyScope), context.parserOptions.ecmaVersion);
```

Comment:
I think this is not right. 

```js
			for (let i = 0; i < arr.length; i += 1) {
				console.log(arr[1]);
				function foo(element) {
					console.log(element);
				}
			}
```

It should allow use variable `element`.

this `foo` function scope should not included, because it didn't use `arr[1]`, this is why `avoidCapture` didn't recursive automaticlly.

We need collect copes from references.

---
BTW: don't squash commits, this makes review harder.

Reply:
A few questions for you.

1. By `arr[1]`, do you mean `arr[i]`? The rule doesn't report a violation with `arr[1]`.
2. In your example, I see that we technically can use `element` as our new array variable name, but wouldn't that be unnecessarily confusing? The [no-shadow](https://eslint.org/docs/rules/no-shadow) rule suggests this is bad practice.

Reply:
1. Yes, `arr[1]`, sorry.
2. Yes `no-shadow` suggest no variable shadowing, it's default to `function`, do you use `all` option? My example is about function scope, but there are other scopes.

This example

```js
for (let i = 0; i < errors.length; i++) {
  console.log(errors[i]);

  try {
    doSomethingThrowError();
  } catch(error) {}
}
```

After #745 we will fix it to 

```js
for (const error_ of errors) {
  console.log(error_);

  try {
    doSomethingThrowError();
  } catch(error) {}
}
```

I don't want use `error_`, I can use `error`.

Reply:
Anyway, this is not very important, let's keep it this way now.

I'm going to check this PR again.

Reply:
Thanks for the explanation. I will follow-up with your recommendation if I have a chance.